feat(dashboards): allow one insight on many dashboards

pauldambra opened the pull request:
Opening a PR as a mini-RFC

# Dashboards Model relations

## Currently

## At the model level

`Insight`s can be saved by themselves _or_ on a `Dashboard`

If they are saved to a dashboard and the dashboard is deleted the insight is deleted too

The relationship is different depending on which side you start at

```mermaid
erDiagram
    INSIGHT ||--|o DASHBOARD : ""
    DASHBOARD ||--o{ INSIGHT: "items"
```

* A single `Insight` can be on zero or one `Dashboard`
* A single `Dashboard` can have zero to many `Insight`s

The foreign key is specified on `Insight`. The relationship specifies a "related_name" so `Dashboard` has a relationship back to `Insight`

## At the API level

### loading a dashboard

`https://app.posthog.com/api/projects/:team_id/dashboards/:dashboard_id/?refresh=false`

Viewing a dashboard makes an API call for that dashboard by ID

That call returns a structure

```
{
  "id": dashboard_id,
  "items": [
    // list of insights on dashboard
  ],
  "created_at": "2022-01-12T19:39:35.014190Z",
  "created_by": {
    // user
  },
  // .... other fields
}
```

### loading an insight

`https://app.posthog.com/api/projects/:team_id/insights/?short_id=the-short-id`

Viewing an insight directly makes an API call by short id

This call returns a paginated structure (ideally (always?) with one result)

```
{
  "count": 1,
  "next": null,
  "previous": null,
  "results": [
    {
      "short_id": "the-short-id",
      "dashboard": dashboard_id,
        // .... other fields
    }
  ]
}
```

The API maps the same relationship:

* Dashboards have many "items"
* Insights have a single dashboard

### Desired relationship

```mermaid
erDiagram
    INSIGHT }o--o{ DASHBOARD : ""
```

The desired relationship is that:

* Dashboards have zero to many insights
* Any Insight can be on many dashboards 

Both insights and dashboards can exist without the other

Django many-to-many relationships implicitly create a join table

```mermaid
erDiagram
    INSIGHT ||--o{ DASHBOARD_INSIGHT: ""
    DASHBOARD_INSIGHT }o--|| DASHBOARD: ""
```

Manually creating that join model would allow future changes which could store data unique to that dashboard/item relation

E.g. a different date range applied to the same insight on different dashboards 

## Complications

### The insight model is stored in a posthog_dashboarditem table

Dashboard Item might not be the correct name for the relation. Dashboards may in future hold more than only insights. E.g. Markdown blocks

Leaving Insights on "dashboard item" constrains either future naming, or means a markdown block would be modelled as an insight in code.

To change this we would need a migration to rename the table or to copy the data to a new table

That migration would need to work for cloud and self-host

### the insight model already has a one way relationship direct to dashboard

Any migration to a join table would need to migrate the items relationship off Insight and keep any existing relationship

## Proposed Approach

Ultimately, the approach is to add a new relation onto `Dashboard`

`insights = models.ManyToManyField(Insight)`

and deprecate the existing `dashboard` field on `Insight`

This allows querying in both directions. And can have a manual "through" table specified in case future changes need to store data associated with a particular insight/dashboard combo. E.g. different date filters on the "last month" dashboard and "this week" dashboard 

#### (optional) Rename the posthog_dashboarditems table

[PostgreSQL documentation on alter table](https://www.postgresql.org/docs/current/sql-altertable.html) for rename says: 

> "There is no effect on the stored data."

It would need testing, but this suggests we can have a single migration and code change that renames the insight table and changes the table name in the model

**TODO** This could lead to a DB that has had the table renamed and running instances that are expecting the old table name. That would need to be addressed, adding complication to this work

**It is _quicker_ to leave Insights on this table until PostHog 2.0 allows breaking changes and add a TODO to that effect on the model** 

#### change the Insight API to present an array of dashboard ids

Currently, the API presents a single dashboard ID. The UI uses this to provide a button to jump to the dashboard. This would need to change to a dropdown button showing the one to many dashboards you could visit from a given insight 

The following UI interactions would need to change:

* adding to a dashboard becomes multiselect
* showing what dashboard an insight is on can't be a single button

![2022-04-04 14 42 30](https://user-images.githubusercontent.com/984817/161557524-11b44b2a-4032-4da0-8c5d-e7a30f80d818.gif)

There are 17 places that the UI code knows that dashboard ID is `number | null` and changing to `number | number[]| null`  or adding a new `number[]` property would need to be accounted for


#### Put a new relation on dashboards to a new DashboardItem model

`insights = models.ManyToManyField(Insight)`

Initially empty, the dashboard API would read from both `insights` and `items` relations and union and deduplicate the results of this relation and the existing `Items` relation

#### 3. Migrate existing dashboard-insight relation onto the new model

In a perfect world this _could_ be as simple as looping through `Insight` models, prefetching their `dashboard`, and storing the relation onto the new `insights` relation on that instance of `Dashboard`

Based on the number of Insights stored on cloud. This would take more than long enough to cause deployments to fail or data to appear to be inconsistent.

~~The join migration would need to be manually written as a fast set based SQL operation~~ [see comment](https://github.com/PostHog/posthog/pull/9331#issuecomment-1088534414)

Or the `Insight` model would union and deduplicate its old and new relation and present that to the outside world. While an async migration paged through the DB updates needed

pauldambra committed:
add a TODO on insight

pauldambra committed:
add manytomany dashboard to insight relation

pauldambra committed:
merge from master and update migration

pauldambra committed:
remove unnecessary change to IDE config file

Twixes commented:
Yeah, we'll definitely need a reworked modal to go from a boolean approach (either is on a dashboard or isn't) to a number (anywhere between 0 and n dashboards).
One tricky thing we definitely need to keep in mind in this project is dashboard permissions: currently when an insight belongs to a dashboard the user can't edit, that user also can't edit that insight. But what if only one out of a few dashboards the insight is on is restricted (e.g. user can edit "Funnels usage" but can't edit "Board meeting metrics")? 🤔

pauldambra committed:
update migration to add unique constraint on posthog_dashboardinsight joins

pauldambra committed:
Merge branch 'master' into many-to-many-insights

pauldambra commented:
Added many insights using

```
insert into posthog_dashboarditem (short_id, team_id, filters, favorited, is_sample, deleted, saved, updated_at,
                                   layouts, last_modified_at, refreshing, dashboard_id)
select left(md5((i::int+3000000)::text), 12),
       1,
       to_jsonb('{}'::text),
       False,
       False,
       False,
       False,
       now(),
       to_jsonb('{}'::text),
       now(),
       False,
       1
from generate_series(1, 1000000) s(i);
```

and inserted them in one step as a _synchronous_ migration

```
insert into posthog_dashboardinsight (insight_id, dashboard_id)
select id as insight_id, dashboard_id from posthog_dashboarditem
where dashboard_id is not null
ON CONFLICT ON CONSTRAINT posthog_dashboardinsight_dashboard_id_insight_id_675423e4_uniq
DO NOTHING;
```

This took between 18 and 25 seconds over three attempts

A million is **a lot** of insights but this demonstrates that a single insert migration _could_ timeout and cause issues with upgrades

Twixes commented:
That's my main worry with splitting `posthog_insight` out of `posthog_dashboarditem`… It shouldn't be an issue for self-hosted setups, but Cloud is tricky. Chunking the migration into more doable portions may help, but if that still takes more than 10 seconds, I worry about the impact of (and on) users editing insights at that time. @guidoiaquinti may have some more/better tips here.

guidoiaquinti commented:
> That's my main worry with splitting `posthog_insight` out of `posthog_dashboarditem`… It shouldn't be an issue for self-hosted setups, but Cloud is tricky. Chunking the migration into more doable portions may help, but if that still takes more than 10 seconds, I worry about the impact of (and on) users editing insights at that time. @guidoiaquinti may have some more/better tips here.

That's a valid concern. As we don't know what's our _customer p100 sort by row count_ on self-hosted, we should make this migration idempotent and split it into chunks in a loop (e.g. 100 rows/query) that will be retried in case of failures. 

There's also a (probably marginal) race condition:

t0 - app is running
t1 - user starts to upgrade the app -> migration starts, web is still running the old version
t2 - user adds an insight (as the app is the old version we sill write the record to the "old" table)
t3 - migration completes without migrating the insight
t4 - app migration completes, web is now running the new version
t5 - there's an orphan record

pauldambra committed:
Add migration that pages through old dashboard item relation and copies to new dashboard insight relation

pauldambra committed:
correct reverse relationship name

pauldambra committed:
hide the insight/item split from insight patch calls

pauldambra committed:
creating an insight with a dashboard id adds the dashboard to the manytomany relation

pauldambra committed:
make the new dashboard tests less rubbish

pauldambra committed:
remove second migration

pauldambra committed:
don't hardcode ids in tests

pauldambra committed:
don't hardcode ids in tests

pauldambra committed:
can safely add dashboard links on create and update

pauldambra committed:
write a test to show that loading a dashboard is nplus1

pauldambra committed:
make test pass and add a comment that adding insights to dashboards _is_ n+1

pauldambra committed:
dashboard does not show soft deleted insights

pauldambra committed:
skip test instead of having incorrect assertion

clarkus commented:
I made [some progress](https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=8803%3A44744) from the insights side of things. This allows for adding an insight to one or more dashboards, or creating a new dashboard and then adding. I am working through the dashboard side of things next. 

![Share Modal - single share link](https://user-images.githubusercontent.com/254612/161866696-c75e22bd-00d5-4ad7-8452-132a2fd123b2.png)
![Group 8972](https://user-images.githubusercontent.com/254612/161866699-a3205bf9-9f1a-4a96-9dbb-790355ccf555.png)
<img width="684" alt="Screen Shot 2022-04-05 at 4 22 24 PM" src="https://user-images.githubusercontent.com/254612/161866755-dd9b64d8-f94d-41bd-acb6-bf3217819c1d.png">

pauldambra committed:
start refactoring insight tests

pauldambra committed:
Merge branch 'master' into many-to-many-insights

pauldambra committed:
more test refactoring and reduce insight loading from 13 to 9 queries

pauldambra committed:
move all insight tests to creating insights through the API not the ORM

pauldambra committed:
use splats not pipes because mypy

Twixes commented on posthog/api/insight.py:
```diff
@@ -270,7 +268,15 @@ def get_serializer_class(self) -> Type[serializers.BaseSerializer]:
+        queryset = queryset.select_related("last_modified_by")
+        queryset = queryset.select_related("team")
+        queryset = queryset.select_related("dashboard__team")
+        queryset = queryset.select_related("dashboard__team__organization")
```
Protip: you do `queryset = super().get_queryset().prefetch_related("foo", "foo_xyz").select_related("bar", "baz")` to amke this more concise ;)

hazzadous commented on posthog/models/dashboard.py:
```diff
@@ -104,3 +106,19 @@ def get_analytics_metadata(self) -> Dict[str, Any]:
+class DashboardInsight(models.Model):
+    """
+    An explicit join model for adding insights to dashboards allows for future extension of the relationship
+    E.g. allowing the same insight to have different filters on different dashboards
```
I guess the more immediate one is that `Insight.layouts` needs to be associated with a specific `Dashboard`. We could either put it here, or as a `Dashboard.layouts`. Is this something that will be handled in this PR or a followup?

pauldambra committed:
collapse prefetch and select related

pauldambra committed:
dashboard permissions propogated through new relation

pauldambra committed:
check permissions on old and new relation

pauldambra committed:
no longer makes sense for insight basic serializer to try and grab filters from a single dashboard

pauldambra committed:
satisfy typechecker

pauldambra commented:
TO-DO

- [x] add new many-to-many relation
- [x] hide insight/item split from read side
- [x] hide insight/item split from write side
- [x] add migration of existing relation into new one
- [x] add new dashboards property on insight API
- [ ] migrate FE data onto new dashboards property
- [ ] if someone's access to an insight is read-only on any dashboard, it is read-only everywhere
- [ ] update UI so insights can be added to multiple dashboards
- [ ] update filter based hash key to cope with more than one dashboard
- [ ] update insight caching to cope with multiple dashboards

pauldambra committed:
deal with caching WIP

pauldambra committed:
a pass at dashboard+insight specific filter hashes

pauldambra committed:
wip

pauldambra committed:
restrict but don't flatten the number of queries to load a dashboard after connecting dashboards and insights

pauldambra committed:
correctly dedupe and update insights

pauldambra committed:
fix cache test

pauldambra committed:
raise validation error on illegal dashboard link

pauldambra committed:
make insight to representation dashboard aware filtering work

pauldambra committed:
team id in test needs to be consistent

pauldambra committed:
fix insight patching behaviour

pauldambra committed:
better linking between dashboard and insight

pauldambra committed:
convert to many-to-many without manual through table - WIP

pauldambra committed:
works but never writes to the cache

pauldambra committed:
resolve nplus1 test

pauldambra committed:
cache key WIP

pauldambra committed:
WIP

pauldambra committed:
improvements to keeping dashboard and dashboards in sync

pauldambra committed:
add comment with todo to update_cache

pauldambra committed:
remove unnecessary todo from dashboard logic

neilkakkar commented:
Q: Is there a link to somewhere I can see where we decided doing this makes sense / solves a good use case / etc. etc. ? This mini-RFC feels more implementation focused, and I'm missing context 😅

mariusandra commented on posthog/test/test_dashboard_insight_model.py:
```diff
@@ -0,0 +1,29 @@
+from posthog.test.base import BaseTest
+
+
+class TestDashboardInsightModel(BaseTest):
```
I don't see the `DashboardInsight` itself. Should this be the table between dashboards and insights? If so, is this the right name for such a table, given it'll contain 1) what to show 2) coordinates of the item on the dashboard, 3) color of the item. The "item" is currently just an insight, but it can also be a block of text, the events list table, or whatever. Should we make a new many-to-many model for every new type of dashboard element?

mariusandra reviewed (commented):
I think this is probably the least disruptive way to bring about this database change 👍. 

This is still a draft, so I'm not sure what is in the TODO list and what is not, nevertheless here and inline are some comments/questions. Happy to address them myself as well.

1. @pauldambra when you did your [test here](https://github.com/PostHog/posthog/pull/9331#issuecomment-1088534414),... did the table you were inserting to have an index or not? Those 18-25 seconds for copying 1M rows from one table to another seem like a lot. Could it be that this was the case, and this insert could be made much faster by cleverly dropping and creating indexes? This feels way too long.

2. We already need an explicit table for the relation, as the fields `layouts` and `color` should go there, and be removed from `insight`. They can be unique for each dashboards.

pauldambra commented:
superceded by #9416

pauldambra commented:
@neilkakkar I missed your comment, sorry

There's not much more there but the RFC for the insights work is here https://github.com/PostHog/product-internal/blob/5c4efc9f46416dabd239a5ef931613e035596a18/requests-for-comments/2022-03-14%20nail%20insights.md
